### ISSUE TITLE
User's news feed aggregation

### ISSUE THREAD

**Issue by fuksi** [loklak/loklak_server#94]
Currently, the only way to get a user's news feed is straight from twitter API with user authorization.
As far as I know, a user's timeline page is not available to scrape (it's only available to that user). What if we can create a work around to clone the service as well?

The original thought/approach is: 
- Make a request to account servlet for the list of following
- For each of the following, make a request to search servlet for the latest tweet
- Sort all results from newest to latest

The problem with this approach is:
- That's a lot of http request and also search jobs for the server
- The delays
- Sorting in client in relative slow

What if the server can provide aggregation search for a timeline, e.g. `?newsfeed.json&screen_name=auser` . This request should be accessed from only localhost.

**Comment by Orbiter** [comment]
The users timeline is available with the search request `from:<user>`.
i.e. http://loklak.org/api/search.json?timezoneOffset=-120&q=from%3Amariobehling

**Comment by mariobehling** [comment]
@Orbiter The question is not about all the tweets of one user, the question is how can we get all the tweets of the accounts the users follows. @fuksi's approach seems to be solving this.

**Comment by Orbiter** [comment]
> @Orbiter The question is not about all the tweets of one user, the question is how can we get all the tweets of the accounts the users follows.

yes, sorry for misunderstanding the feature.

> @fuksi's approach seems to be solving this.

well it does not solve this, it just shows up the complexity of the task.

When finding an appropriate solution to this request, I ask myself to solve these details:
- (1) what is the appropriate api within loklak?
- (2) what kind of architecture is appropriate to find a solution if we rely only on our own data (like: loklak does not any more rely on twitter accounts and their data)
- (3) how can be get the timeline from twitter without using another twitter API like https://dev.twitter.com/rest/reference/get/statuses/home_timeline

To make things simple for loklak_webclient, the answer to (1) should be, that the timeline should be returned by the /api/search.json servlet because this servlet returns the same data type (tweets) as expected for the timeline.

The complexity to solve (2) is high: there would be two solutions for this: either make single search requests to the elasticsearch index for each 'following'-enries of the user which could result in thousands of queries which is not feasibe. Or we just iterate through the whole index timeline and collect tweets matching the 'following'-entries of the user, collecting them in the result timeline. This looks feasible but can still cause high IO load to iterate through the index. To enhance this procedure, there could be a sliding cache of the latest X tweets (like all tweets of one day), but that cache would have to be implemented as well. However 'iterating all tweets until <max-tweets> following-tweets are collected' is most feasible.

The logical view on this feature is: its simply a filter and can be considered as constraint. Therefore the syntaxt to query all the tweets which a user is following should be `/following=<user>`. Very simple, and implementing this as a search constraint will also allow to search within all the tweets of accounts the user is following. That should be also applicable applicable to IoT: lets consider that freifunk-routes have statuses stored in the account `$freifunk` which the user is following and such devices have the status message 'fail' if something is wrong. Then searching these statuses would be simply `fail /following=<user>`

To solve (3) we should not depend on another twitter api. Instead, we could use the search scraper and create a combined search query for all accounts the user is following. This could result in a search request to twitter like `from:<user1> OR from:<user2> OR ...` but this query string could have a limitation We must find out this limitation.

---
tl;dr
I will implement this as search with a constraint, just search for `/following=<user>` at `/api/search.json` to get all tweets which the user is following.

**Comment by smokingwheels** [comment]
Could you call the mobile site of twitter in the API eg https://mobile.twitter.com/smokingwheels ?

**Comment by Orbiter** [comment]
>Could you call the mobile site of twitter in the API

Thats a great idea for the scraper!

**Comment by mariobehling** [comment]
@Orbiter This would still be a great feature, but do you think this should actually be on the list to be implemented?
